### ISSUE TITLE
jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unsupported attribute type for attribute: array

### ISSUE THREAD

**Issue by mariogeiger** [jax-ml/jax#24160]
## bug
I have a bug when I try to pass an array as attribute to the XLA/FFI interface.

## reproduce

content of `foo.cc`
```c++
#include "xla/ffi/api/c_api.h"
#include "xla/ffi/api/ffi.h"

namespace ffi = xla::ffi;

ffi::Error foo_host(ffi::Span<const int64_t> array,
                    ffi::Buffer<ffi::DataType::F32> input,
                    ffi::Result<ffi::Buffer<ffi::DataType::F32>> output)
{
  return ffi::Error::Success();
}

XLA_FFI_DEFINE_HANDLER_SYMBOL(
  foo,
  foo_host,
  ffi::Ffi::Bind()
    .Attr<ffi::Span<const int64_t>>("array")
    .Arg<ffi::Buffer<ffi::DataType::F32>>()  // input
    .Ret<ffi::Buffer<ffi::DataType::F32>>()  // output
);
```
```
# g++ -std=c++17 -fPIC -I /opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include -shared -o libfoo.so foo.cc
```

<details><summary>Full output</summary>
<p>

``` 
In file included from foo.cc:2:
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/ffi.h:626:68: warning: ‘always_inline’ function might not be inlinable [-Wattributes]
  626 | XLA_FFI_ATTRIBUTE_ALWAYS_INLINE std::optional<Buffer<dtype, rank>> DecodeBuffer(
      |                                                                    ^~~~~~~~~~~~
In file included from /opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/ffi.h:48:
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h: In function ‘std::ostream& operator<<(std::ostream&, XLA_FFI_ExecutionStage)’:
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h:172:1: warning: control reaches end of non-void function [-Wreturn-type]
  172 | }
      | ^
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h: In function ‘std::ostream& operator<<(std::ostream&, XLA_FFI_AttrType)’:
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h:158:1: warning: control reaches end of non-void function [-Wreturn-type]
  158 | }
      | ^
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h: In function ‘std::ostream& operator<<(std::ostream&, XLA_FFI_DataType)’:
/opt/python/cp310-cp310/lib/python3.10/site-packages/jaxlib/include/xla/ffi/api/api.h:145:1: warning: control reaches end of non-void function [-Wreturn-type]
  145 | }
      | ^
```

</p>
</details> 

content of `foo.py`
```python
import ctypes

import jax
import jax.numpy as jnp
import numpy as np
from jax.extend import ffi
from jax.lib import xla_client
import os

library = ctypes.cdll.LoadLibrary(os.path.join(os.path.dirname(__file__), "libfoo.so"))

xla_client.register_custom_call_target(
    name="foo",
    fn=ffi.pycapsule(getattr(library, "foo")),
    platform="CUDA",
    api_version=1,
)

ffi.ffi_call(
    "foo",
    (jax.ShapeDtypeStruct((2, 2, 2), jnp.float32),),
    jnp.zeros((2, 2, 2), dtype=jnp.float32),
    array=np.array([14, 42, 42], dtype=np.int64),
)
```

```
# python foo.py
[see full output below]
jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unsupported attribute type for attribute: array
```

<details><summary>Full output</summary>
<p>

```
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/bug/foo.py", line 19, in <module>
    ffi.ffi_call(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/extend/ffi.py", line 247, in ffi_call
    results = ffi_call_p.bind(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 438, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 442, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 948, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/dispatch.py", line 90, in apply_primitive
    outs = fun(*args)
jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unsupported attribute type for attribute: array
[root@db1ac34-lcedt bug]# JAX_TRACEBACK_FILTERING=off python foo.py
Traceback (most recent call last):
  File "/home/bug/foo.py", line 19, in <module>
    ffi.ffi_call(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/extend/ffi.py", line 247, in ffi_call
    results = ffi_call_p.bind(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 438, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 442, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 948, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/dispatch.py", line 90, in apply_primitive
    outs = fun(*args)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 180, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/pjit.py", line 356, in cache_miss
    outs, out_flat, out_tree, args_flat, jaxpr, attrs_tracked = _python_pjit_helper(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/pjit.py", line 189, in _python_pjit_helper
    out_flat = pjit_p.bind(*args_flat, **p.params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 2781, in bind
    return self.bind_with_trace(top_trace, args, params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 442, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/core.py", line 948, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/pjit.py", line 1764, in _pjit_call_impl
    return xc._xla.pjit(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/pjit.py", line 1739, in call_impl_cache_miss
    out_flat, compiled = _pjit_call_impl_python(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/pjit.py", line 1669, in _pjit_call_impl_python
    ).compile(compile_options)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2315, in compile
    executable = UnloadedMeshExecutable.from_hlo(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2829, in from_hlo
    xla_executable = _cached_compilation(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2641, in _cached_compilation
    xla_executable = compiler.compile_or_get_cached(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 427, in compile_or_get_cached
    return _compile_and_write_cache(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 655, in _compile_and_write_cache
    executable = backend_compile(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/profiler.py", line 333, in wrapper
    return func(*args, **kwargs)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 273, in backend_compile
    raise e
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 267, in backend_compile
    return backend.compile(built_c, compile_options=options)
jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unsupported attribute type for attribute: array
```

</p>
</details> 

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.34
jaxlib: 0.4.34
numpy:  2.1.2
python: 3.10.14 (main, Jul 15 2024, 05:45:40) [GCC 12.2.1 20221121 (Red Hat 12.2.1-7)]
jax.devices (2 total, 2 local): [CudaDevice(id=0) CudaDevice(id=1)]
process_count: 1
platform: uname_result(system='Linux', node='db1ac34-lcedt', release='6.8.0-40-generic', version='#40~22.04.3-Ubuntu SMP PREEMPT_DYNAMIC Tue Jul 30 17:30:19 UTC 2', machine='x86_64')


$ nvidia-smi
Mon Oct  7 13:59:01 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.28.03              Driver Version: 560.28.03      CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA RTX 5880 Ada Gene...    Off |   00000000:01:00.0 Off |                  Off |
| 30%   41C    P0             27W /  285W |     346MiB /  49140MiB |      1%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA RTX 5880 Ada Gene...    Off |   00000000:09:00.0 Off |                  Off |
| 30%   37C    P0             22W /  285W |     346MiB /  49140MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
+-----------------------------------------------------------------------------------------+
```

**Comment by mariogeiger** [comment]
```python
import ctypes

import jax
import jax.numpy as jnp
import numpy as np
from jax.extend import ffi
from jax.lib import xla_client
import os

print(f"jax version: {jax.__version__}")
# jax version: 0.4.34

library = ctypes.cdll.LoadLibrary(os.path.join(os.path.dirname(__file__), "libfoo.so"))

xla_client.register_custom_call_target(
    name="foo",
    fn=ffi.pycapsule(getattr(library, "foo")),
    platform="cpu",
    api_version=1,
)


def fun(x):
    ffi.ffi_call("foo", (x,), x, array=np.array([14, 42, 42], dtype=np.int64))


print(jax.make_jaxpr(fun)(jnp.zeros((2, 2, 2), dtype=jnp.float32)))
"""
{ lambda ; a:f32[2,2,2]. let
    _:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in () }
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() {
    return
  }
}

HloModule jit_fun, is_scheduled=true, entry_computation_layout={()->()}, frontend_attributes={fingerprint_before_lhs="489c8f99adcd8c8f07fa8a7d0f5ef40e"}

ENTRY %main.2 () -> () {
  ROOT %tuple.1.0 = () tuple()
}


[root@db1ac34-lcedt bug]# python foo.py
jax version: 0.4.34
{ lambda ; a:f32[2,2,2]. let
    _:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in () }
"""

print(jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).as_text())
"""
{ lambda ; a:f32[2,2,2]. let
    _:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in () }
[root@db1ac34-lcedt bug]# python foo.py
jax version: 0.4.34
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() {
    return
  }
}

"""

print(jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).compile().as_text())
"""
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() {
    return
  }
}

[root@db1ac34-lcedt bug]# python foo.py
jax version: 0.4.34
HloModule jit_fun, is_scheduled=true, entry_computation_layout={()->()}, frontend_attributes={fingerprint_before_lhs="489c8f99adcd8c8f07fa8a7d0f5ef40e"}

ENTRY %main.2 () -> () {
  ROOT %tuple.1.0 = () tuple()
}

"""
```

**Comment by dfm** [comment]
Whoops! We need a return inside of `fun`:

```python
def fun(x):
  return ffi.ffi_call("foo", (x,), x, array=np.array([14, 42, 42], dtype=np.int64))
```

**Comment by mariogeiger** [comment]
```
# python foo.py
jax version: 0.4.34
{ lambda ; a:f32[2,2,2]. let
    _:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in () }
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() {
    return
  }
}

HloModule jit_fun, is_scheduled=true, entry_computation_layout={()->()}, frontend_attributes={fingerprint_before_lhs="489c8f99adcd8c8f07fa8a7d0f5ef40e"}

ENTRY %main.2 () -> () {
  ROOT %tuple.1.0 = () tuple()
}


[root@db1ac34-lcedt bug]# python foo.py
jax version: 0.4.34
{ lambda ; a:f32[2,2,2]. let
    b:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in (b,) }
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2x2x2xf32> {mhlo.layout_mode = "default"}) -> (tensor<2x2x2xf32> {jax.result_info = "[0]", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.custom_call @foo(%arg0) {backend_config = "", mhlo.backend_config = {array = array<i64: 14, 42, 42>}, operand_layouts = [dense<[2, 1, 0]> : tensor<3xindex>], result_layouts = [dense<[2, 1, 0]> : tensor<3xindex>]} : (tensor<2x2x2xf32>) -> tensor<2x2x2xf32>
    return %0 : tensor<2x2x2xf32>
  }
}

Traceback (most recent call last):
  File "/home/bug/foo.py", line 49, in <module>
    print(jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).compile().as_text())
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/stages.py", line 660, in compile
    self._lowering.compile(**kw),  # pytype: disable=wrong-keyword-args
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2315, in compile
    executable = UnloadedMeshExecutable.from_hlo(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2829, in from_hlo
    xla_executable = _cached_compilation(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2641, in _cached_compilation
    xla_executable = compiler.compile_or_get_cached(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 427, in compile_or_get_cached
    return _compile_and_write_cache(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 655, in _compile_and_write_cache
    executable = backend_compile(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/profiler.py", line 333, in wrapper
    return func(*args, **kwargs)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 273, in backend_compile
    raise e
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 267, in backend_compile
    return backend.compile(built_c, compile_options=options)
jaxlib.xla_extension.XlaRuntimeError: UNIMPLEMENTED: No registered implementation for custom call to foo for platform CUDA
```

**Comment by dfm** [comment]
Thanks! The HLO looks right, and now you're getting a different error. Are you still registering `foo`?

**Comment by mariogeiger** [comment]
very sorry, while trying to debug I tried to replace "CUDA" by "cpu".
Here is the output when I put back "CUDA" as in the original code I shared. 

```
# python foo.py
jax version: 0.4.34
{ lambda ; a:f32[2,2,2]. let
    b:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in (b,) }
module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2x2x2xf32> {mhlo.layout_mode = "default"}) -> (tensor<2x2x2xf32> {jax.result_info = "[0]", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.custom_call @foo(%arg0) {backend_config = "", mhlo.backend_config = {array = array<i64: 14, 42, 42>}, operand_layouts = [dense<[2, 1, 0]> : tensor<3xindex>], result_layouts = [dense<[2, 1, 0]> : tensor<3xindex>]} : (tensor<2x2x2xf32>) -> tensor<2x2x2xf32>
    return %0 : tensor<2x2x2xf32>
  }
}

Traceback (most recent call last):
  File "/home/bug/foo.py", line 49, in <module>
    print(jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).compile().as_text())
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/stages.py", line 660, in compile
    self._lowering.compile(**kw),  # pytype: disable=wrong-keyword-args
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2315, in compile
    executable = UnloadedMeshExecutable.from_hlo(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2829, in from_hlo
    xla_executable = _cached_compilation(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 2641, in _cached_compilation
    xla_executable = compiler.compile_or_get_cached(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 427, in compile_or_get_cached
    return _compile_and_write_cache(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 655, in _compile_and_write_cache
    executable = backend_compile(
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/profiler.py", line 333, in wrapper
    return func(*args, **kwargs)
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 273, in backend_compile
    raise e
  File "/opt/python/cp310-cp310/lib/python3.10/site-packages/jax/_src/compiler.py", line 267, in backend_compile
    return backend.compile(built_c, compile_options=options)
jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unsupported attribute type for attribute: array
```

**Comment by dfm** [comment]
One last thing to try on your end: Can you run all of this on CPU (e.g. set the environment variable `JAX_PLATFORM_NAME=cpu`, and register the FFI handler on "cpu" instead of CUDA) to see if you see the same behavior?

**Comment by mariogeiger** [comment]
```python
import ctypes

import jax
import jax.numpy as jnp
import numpy as np
from jax.extend import ffi
from jax.lib import xla_client
import os

library = ctypes.cdll.LoadLibrary(os.path.join(os.path.dirname(__file__), "libfoo.so"))

xla_client.register_custom_call_target(
    name="foo",
    fn=ffi.pycapsule(getattr(library, "foo")),
    platform=os.environ.get("JAX_PLATFORM_NAME", "cpu"),
    api_version=1,
)


def fun(x):
    return ffi.ffi_call("foo", (x,), x, array=np.array([14, 42, 42], dtype=np.int64))


print(f"{jax.__version__=}")
print(f"{jax.make_jaxpr(fun)(jnp.zeros((2, 2, 2), dtype=jnp.float32))=}")
print(f"{jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).as_text()=}")
print(
    f"{jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).compile().as_text()=}"
)
```

```
# JAX_PLATFORM_NAME=cpu python foo.py
jax.__version__='0.4.34'
jax.make_jaxpr(fun)(jnp.zeros((2, 2, 2), dtype=jnp.float32))={ lambda ; a:f32[2,2,2]. let
    b:f32[2,2,2] = ffi_call[
      array=HashableArray([14 42 42])
      has_side_effect=False
      result_avals=(ShapedArray(float32[2,2,2]),)
      target_name=foo
      vectorized=False
    ] a
  in (b,) }
jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).as_text()='module @jit_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {\n  func.func public @main(%arg0: tensor<2x2x2xf32> {mhlo.layout_mode = "default"}) -> (tensor<2x2x2xf32> {jax.result_info = "[0]", mhlo.layout_mode = "default"}) {\n    %0 = stablehlo.custom_call @foo(%arg0) {backend_config = "", mhlo.backend_config = {array = array<i64: 14, 42, 42>}, operand_layouts = [dense<[2, 1, 0]> : tensor<3xindex>], result_layouts = [dense<[2, 1, 0]> : tensor<3xindex>]} : (tensor<2x2x2xf32>) -> tensor<2x2x2xf32>\n    return %0 : tensor<2x2x2xf32>\n  }\n}\n'
jax.jit(fun).lower(jnp.zeros((2, 2, 2), dtype=jnp.float32)).compile().as_text()='HloModule jit_fun, is_scheduled=true, entry_computation_layout={(f32[2,2,2]{2,1,0})->f32[2,2,2]{2,1,0}}, allow_spmd_sharding_propagation_to_parameters={true}, allow_spmd_sharding_propagation_to_output={true}\n\nENTRY %main.3 (Arg_0.1: f32[2,2,2]) -> f32[2,2,2] {\n  %Arg_0.1 = f32[2,2,2]{2,1,0} parameter(0), metadata={op_name="x"}\n  ROOT %custom-call.2 = f32[2,2,2]{2,1,0} custom-call(f32[2,2,2]{2,1,0} %Arg_0.1), custom_call_target="foo", operand_layout_constraints={f32[2,2,2]{2,1,0}}, api_version=API_VERSION_TYPED_FFI, metadata={op_name="jit(fun)/jit(main)/ffi_call" source_file="/home/bug/foo.py" source_line=21}, backend_config={array = array<i64: 14, 42, 42>}\n}\n\n'
```

**Comment by dfm** [comment]
Interesting! So there's a mismatch between the CPU backend and the plugin. What do you see if you run:

```python
from importlib.metadata import version
version("jax-cuda12-plugin")
```

in your environment?

**Comment by dfm** [comment]
Actually - I can reproduce the issue even with the latest plugin. I can dig into it over here now!

**Comment by dfm** [comment]
I got to the bottom of this. The tl;dr is that because the GPU plugin serializes via VHLO, which [has limited support for array attributes](https://github.com/openxla/stablehlo/issues/2121), we were ending up with the wrong attribute types by the time we reached the FFI attribute decoding. I was able to work around this by [adding support for decoding the appropriate types in XLA](https://github.com/openxla/xla/pull/17981), which has now landed. In a day or two, the [nightly version of jaxlib](https://jax.readthedocs.io/en/latest/installation.html#jax-nightly-installation) will support array arguments via the plugin!
